### PR TITLE
WIP: New output layer

### PR DESCRIPTION
**Pull request by prashant-jayan21**

### CODE REVIEW

File: bi-att-flow-dev/basic/model_nqa.py [hunk 1]
```diff
@@ -183,114 +203,99 @@ def _build_forward(self):
...
-                p0 = attention_layer(config, self.is_train, h, u, h_mask=self.x_mask, u_mask=self.q_mask, scope="p0", tensor_dict=self.tensor_dict)
+                p0 = attention_layer(config, self.is_train, h, u, h_mask=self.x_mask, u_mask=self.q_mask, scope="p0", tensor_dict=self.tensor_dict) # p0 seems to be G in paper
                 first_cell_fw = d_cell2_fw
                 second_cell_fw = d_cell3_fw
                 first_cell_bw = d_cell2_bw
                 second_cell_bw = d_cell3_bw
 
             (fw_g0, bw_g0), _ = bidirectional_dynamic_rnn(first_cell_fw, first_cell_bw, p0, x_len, dtype='float', scope='g0')  # [N, M, JX, 2d]
             g0 = tf.concat(axis=3, values=[fw_g0, bw_g0])
-            (fw_g1, bw_g1), _ = bidirectional_dynamic_rnn(second_cell_fw, second_cell_bw, g0, x_len, dtype='float', scope='g1')  # [N, M, JX, 2d]
```
Comment (dhruv100691) [inline bi-att-flow-dev/basic/model_nqa.py]: dont delete this, this is the second bidirectional layer of modelling layer

File: bi-att-flow-dev/basic/model_nqa.py [hunk 2]
```diff
@@ -183,114 +203,99 @@ def _build_forward(self):
...
+            # print(tf.shape(my_fw_final_state)) # Tensor("model_0/main/Shape_14:0", shape=(3,), dtype=int32, device=/device:GPU:0)
+            # print(my_fw_final_state) # LSTMStateTuple(c=<tf.Tensor 'model_0/main/g1/fw/fw/while/Exit_2:0' shape=(?, 100) dtype=float32>, h=<tf.Tensor 'model_0/main/g1/fw/fw/while/Exit_3:0' shape=(?, 100) dtype=float32>)
+
+            # For now go with my_fw_final_state only
+
+            # Decoder embedding
+
+            # Embedding decoder/matrix
+
+            tgt_vocab_size = VW # hparam # FIXME: Obtain embeddings differently?
```
Comment (dhruv100691) [inline bi-att-flow-dev/basic/model_nqa.py]: total vocab size is not VW, its going to be the length of word_emb_mat

File: bi-att-flow-dev/basic/model_nqa.py [hunk 3]
```diff
@@ -183,114 +203,99 @@ def _build_forward(self):
...
+            print(decoder_emb_inp)
+
+            def decode(helper, scope, reuse=None, maximum_iterations=None):
+                with tf.variable_scope(scope, reuse=reuse):
+                    decoder_cell = BasicLSTMCell(d, state_is_tuple=True) # hparam
+                    projection_layer = layers_core.Dense(
+                        tgt_vocab_size, use_bias=False) # hparam
+
+                    decoder = tf.contrib.seq2seq.BasicDecoder(
+                        decoder_cell, helper, my_fw_final_state,
```
Comment (dhruv100691) [inline bi-att-flow-dev/basic/model_nqa.py]: we can directly pass g1 here..i think. instead of my_fw_state

File: bi-att-flow-dev/basic/model_nqa.py [hunk 4]
```diff
@@ -183,114 +203,99 @@ def _build_forward(self):
...
+                        output_layer=projection_layer) # decoder
+
+                    final_outputs, final_decoder_state, final_sequence_lengths = tf.contrib.seq2seq.dynamic_decode(
+                        decoder, output_time_major=False, impute_finished=True, maximum_iterations=maximum_iterations) # dynamic decoding # TODO: check tutorial for more
+
+                    return final_outputs
+
+            # Decoder
+            training_helper = tf.contrib.seq2seq.TrainingHelper(
```
Comment (dhruv100691) [inline bi-att-flow-dev/basic/model_nqa.py]: lets add the greedy helper also for inference here

File: bi-att-flow-dev/basic/model_nqa.py [hunk 5]
```diff
@@ -183,114 +203,99 @@ def _build_forward(self):
...
+            self.decoder_logits_train = decoder_logits_train
 
     def _build_loss(self):
         config = self.config
         JX = tf.shape(self.x)[2]
         M = tf.shape(self.x)[1]
         JQ = tf.shape(self.q)[1]
+        N = config.batch_size
 
-        loss_mask = tf.reduce_max(tf.cast(self.q_mask, 'float'), 1)
```
Comment (dhruv100691) [inline bi-att-flow-dev/basic/model_nqa.py:250]: mask will be required

File: bi-att-flow-dev/basic/model_nqa.py [hunk 6]
```diff
@@ -183,114 +203,99 @@ def _build_forward(self):
...
-                    self.concat_logits, tf.cast(concat_y, 'float') / tf.reduce_sum(tf.cast(self.wy, 'float')))
-            else:
-                losses = tf.nn.softmax_cross_entropy_with_logits(
-                    self.logits2, tf.cast(tf.reshape(self.wy, [-1, M * JX]), 'float') / tf.reduce_sum(tf.cast(self.wy, 'float')))
-            ce_loss = tf.reduce_mean(loss_mask * losses)
-            """
-            tf.add_to_collection('losses', ce_loss)
+        # Loss
+        crossent = tf.nn.sparse_softmax_cross_entropy_with_logits(labels=self.decoder_targets, logits=self.decoder_logits_train)
+        train_loss = (tf.reduce_sum(crossent * self.target_weights) / tf.cast(N, dtype=tf.float32))
```
Comment (dhruv100691) [inline bi-att-flow-dev/basic/model_nqa.py]: what are these target weights?

File: bi-att-flow-dev/basic/model_nqa.py [hunk 7]
```diff
@@ -183,114 +203,99 @@ def _build_forward(self):
...
-            tf.add_to_collection('losses', ce_loss)
-            tf.add_to_collection("losses", ce_loss2)
-
-        self.loss = tf.add_n(tf.get_collection('losses', scope=self.scope), name='loss')
+        tf.add_to_collection('losses', train_loss)
+
+        self.loss = tf.add_n(tf.get_collection('losses', scope=self.scope), name='loss') # loss for the entire batch ?
         tf.summary.scalar(self.loss.op.name, self.loss)
-        tf.add_to_collection('ema/scalar', self.loss)
+        tf.add_to_collection('ema/scalar', self.loss) # FIXME: Needed?
```
Comment (dhruv100691) [inline bi-att-flow-dev/basic/model_nqa.py]: needed..we should keep it

File: bi-att-flow-dev/basic/trainer.py [hunk 8]
```diff
@@ -46,13 +46,17 @@ def __init__(self, config, models):
...
         grads_list = []
+        print(self.var_list) # TODO: Check is this should not be None?
         for gpu_idx, model in enumerate(models):
             with tf.name_scope("grads_{}".format(gpu_idx)), tf.device("/{}:{}".format(config.device_type, gpu_idx)):
                 loss = model.get_loss()
                 grads = self.opt.compute_gradients(loss, var_list=self.var_list)
                 losses.append(loss)
                 grads_list.append(grads)
 
+        print(losses)
```
Comment (dhruv100691) [inline bi-att-flow-dev/basic/trainer.py]: what is the fix that you made in this file

File: bi-att-flow-dev/basic/model_nqa.py [hunk 9]
```diff
@@ -34,33 +36,51 @@ def __init__(self, config, scope, rep=True):
...
+
+        # Misc
         self.wy = tf.placeholder('bool', [N, None, None], name='wy')
         self.is_train = tf.placeholder('bool', [], name='is_train')
         self.new_emb_mat = tf.placeholder('float', [None, config.word_emb_size], name='new_emb_mat')
         self.na = tf.placeholder('bool', [N], name='na')
 
+        # New placeholders for our case
+        self.decoder_inputs = tf.placeholder('int32', [N, None], name='decoder_inputs') # [batch_size, max words]
+        self.target_sequence_length = tf.placeholder(shape=(N,), dtype=tf.int32, name='target_sequence_length') # batch_size
```
Comment (dhruv100691) [inline bi-att-flow-dev/basic/model_nqa.py:64]: could not find the  code where these are being assigned?

File: bi-att-flow-dev/basic/model_nqa.py [hunk 10]
```diff
@@ -183,114 +203,99 @@ def _build_forward(self):
...
+            embedding_decoder = word_emb_mat
+            # tf.Variable(
+            #     tf.constant(0.0, shape=[tgt_vocab_size, tgt_embedding_size]), trainable=False, name="embedding_decoder"
+            # )
+
+            # Look up embedding
+            decoder_emb_inp = tf.nn.embedding_lookup(embedding_decoder, self.decoder_inputs) # [batch_size, max words, embedding_size]
+            print(decoder_emb_inp)
+
+            def decode(helper, scope, reuse=None, maximum_iterations=None):
```
Comment (dhruv100691) [inline bi-att-flow-dev/basic/model_nqa.py]: i would suggest that instead of using a function(decode function here), put the entire decoder code in a separate variable scope

File: bi-att-flow-dev/basic/model_nqa.py [hunk 11]
```diff
@@ -183,114 +203,99 @@ def _build_forward(self):
...
+            # Decoder embedding
+
+            # Embedding decoder/matrix
+
+            tgt_vocab_size = VW # hparam # FIXME: Obtain embeddings differently?
+            print(tgt_vocab_size)
+            tgt_embedding_size = dw # hparam
+            print(tgt_embedding_size)
+
+            embedding_decoder = word_emb_mat
```
Comment (dhruv100691) [inline bi-att-flow-dev/basic/model_nqa.py]: dont use the same word_emb_mat..create a new decoder emb matrix?

File: bi-att-flow-dev/basic/model_nqa.py (hunk 1 above)
Comment (prashant-jayan21) [reply bi-att-flow-dev/basic/model_nqa.py]: Oh yes yes. Ye temporary hai. I'm going to add that back.

File: bi-att-flow-dev/basic/model_nqa.py (hunk 2 above)
Comment (prashant-jayan21) [reply bi-att-flow-dev/basic/model_nqa.py]: Hmm. Is it not the same thing here? Lemme check...

File: bi-att-flow-dev/basic/model_nqa.py (hunk 3 above)
Comment (prashant-jayan21) [reply bi-att-flow-dev/basic/model_nqa.py]: So this is the initial hidden state of the decoder. And that will come from the final hidden state of the modeling layer. g1 is the output of the modeling layer, not final hidden state.

File: bi-att-flow-dev/basic/model_nqa.py (hunk 4 above)
Comment (prashant-jayan21) [reply bi-att-flow-dev/basic/model_nqa.py]: Yep. On my todo list.

File: bi-att-flow-dev/basic/model_nqa.py (hunk 5 above)
Comment (prashant-jayan21) [reply bi-att-flow-dev/basic/model_nqa.py:250]: So there's a new variable called `target_weights` that basically replaces this. So should be fine I guess.

File: bi-att-flow-dev/basic/model_nqa.py (hunk 6 above)
Comment (prashant-jayan21) [reply bi-att-flow-dev/basic/model_nqa.py]: https://github.com/dhruv100691/CS-546--Narrative-QA/pull/1#discussion_r183571688

Comment (prashant-jayan21) [reply bi-att-flow-dev/basic/model_nqa.py]: Used in `_build_loss`

File: bi-att-flow-dev/basic/model_nqa.py (hunk 7 above)
Comment (prashant-jayan21) [reply bi-att-flow-dev/basic/model_nqa.py]: Haan I still don't understand this well. Samjha dena.

File: bi-att-flow-dev/basic/trainer.py (hunk 8 above)
Comment (prashant-jayan21) [reply bi-att-flow-dev/basic/trainer.py]: Oh nothing actually. To fix this I needed to fix the model actually.

File: bi-att-flow-dev/basic/model_nqa.py (hunk 9 above)
Comment (prashant-jayan21) [reply bi-att-flow-dev/basic/model_nqa.py:64]: Haan yahi to karna hai hum dono ko mil ke 😃

File: bi-att-flow-dev/basic/model_nqa.py (hunk 11 above)
Comment (prashant-jayan21) [reply bi-att-flow-dev/basic/model_nqa.py]: But we discussed that it should be the same right?

File: bi-att-flow-dev/basic/model_nqa.py [hunk 12]
```diff
@@ -73,6 +73,9 @@ def __init__(self, config, scope, rep=True):
         self.var_list = None
         self.na_prob = None
         self.decoder_logits_train = None
+        self.translations = None
+        self.tgt_sos_id = 2
+        self.tgt_eos_id = 3
```
Comment (prashant-jayan21) [inline bi-att-flow-dev/basic/model_nqa.py]: @dhruv100691 : SOS id and EOS id should be id's that are not already assigned to any other word in our vocab. Using 2 and 3 seem a bit suspicious. Is this okay?

File: bi-att-flow-dev/basic/model_nqa.py [hunk 13]
```diff
@@ -270,8 +279,12 @@ def _build_loss(self):
...
 
         # Loss
-        crossent = tf.nn.sparse_softmax_cross_entropy_with_logits(labels=self.decoder_targets, logits=self.decoder_logits_train)
-        train_loss = tf.reduce_sum(tf.reduce_sum(crossent, axis=1) * loss_mask) / tf.cast(N, dtype=tf.float32)
+        #TODO: cannot calulate loss for test since max iteration cannot be matched to test length
+        if config.mode == 'train':
+            crossent = tf.nn.sparse_softmax_cross_entropy_with_logits(labels=self.decoder_targets, logits=self.decoder_logits_train)
+            train_loss = tf.reduce_sum(tf.reduce_sum(crossent, axis=1) * loss_mask) / tf.cast(N, dtype=tf.float32)
+        else:
+            train_loss=tf.zeros([1],tf.int32)
```
Comment (prashant-jayan21) [inline bi-att-flow-dev/basic/model_nqa.py]: @dhruv100691 : Why this?

File: bi-att-flow-dev/basic/model_nqa.py (hunk 12 above)
Comment (dhruv100691) [reply bi-att-flow-dev/basic/model_nqa.py]: haan..kyunki rest of the words ka id 4 se start karte hain specifically...pehle 2 ids NULL or UNK hai..baaki do yeh

File: bi-att-flow-dev/basic/model_nqa.py (hunk 13 above)
Comment (dhruv100691) [reply bi-att-flow-dev/basic/model_nqa.py]: test loss calculate nhi ho paa rha because...shapes of logits and labels cannot be matched, since voh maximum iterations greedy helper mein should match test label sequences always,
I didnt try to match it also..since calculating test loss is not that imp..as we already have bleu scores

File: bi-att-flow-dev/basic/model_nqa.py (hunk 12 above)
Comment (prashant-jayan21) [reply bi-att-flow-dev/basic/model_nqa.py]: Oh okay. Cool cool.

File: bi-att-flow-dev/basic/model_nqa.py (hunk 13 above)
Comment (prashant-jayan21) [reply bi-att-flow-dev/basic/model_nqa.py]: Oh true true! Ya bleu scores should suffice.
